### PR TITLE
[qualifier] DSS0210,A2-7-2,7 validate authentication for OIR CRUD+search endpoints

### PR DESCRIPTION
**Pull request by Shastick**
Extend the authentication checking scenario to include the OIR endpoint.

### CODE REVIEW

**Comment by BenjaminPelletier** [comment]
If a PR is not ready for review (because, for instance, it depends on another PR that has not yet been merged), let's mark it as draft since it is not yet ready for review.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py [hunk 1]
```diff
@@ -173,6 +195,44 @@ def _ensure_clean_workspace_step(self):
         self.end_test_step()
 
     def _ensure_test_entities_dont_exist(self):
+
+        # Drop OIR's first: subscriptions may be tied to them and can't be deleted
+        # as long as they exist
+        # TODO cleanly move this into the test fragments once the relevant PRs (notably #535) are merged
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py:201]: Is the plan to wait for merging this PR?

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py [hunk 2]
```diff
@@ -173,6 +195,44 @@ def _ensure_clean_workspace_step(self):
...
+                self.record_query(q)
+            except QueryError as qe:
+                self.record_queries(qe.queries)
+                if qe.queries[0].response.status_code == 404:
+                    return  # All is good
+                else:
+                    query = qe.queries[0]
+                    check.record_failed(
+                        summary=f"Could not query OIR {self._test_id}",
+                        details=f"When attempting to query OIR {self._test_id} from the DSS, received {query.response.status_code}: {qe.msg}",
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py:216]: nit, this should be enough:
```suggestion
                        details=f"When attempting to query OIR {self._test_id} from the DSS, received {query.response.status_code}: {qe}",
```

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md [hunk 3]
```diff
@@ -173,6 +173,138 @@ it is in violation of **[astm.f3548.v21.DSS0210,A2-7-2,7](../../../../../require
 If the DSS does not allow searching for subscriptions when valid credentials are presented,
 it is in violation of **[astm.f3548.v21.DSS0005,5](../../../../../requirements/astm/f3548/v21.md)**.
 
+### Operational intents endpoints authentication test step
+
+#### 🛑 Unauthorized requests return the proper error message body check
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:178]: Looks like this should be a fragment?

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+            self._gen_val.verify_4xx_response(wrong_scope_q)
+
+        valid_q = self._gen_val.query_valid_auth(**query_kwargs)
+        with self._scenario.check(
+            "Mutate operational intent reference with valid credentials", self._pid
+        ) as check:
+            if valid_q.status_code != 200:
+                check.record_failed(
+                    summary=f"Expected 200, got {valid_q.status_code}",
+                    details=f"Mutation is expected to have succeeded, but got status {valid_q.status_code} with body {valid_q.response.json} instead",
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:335]: Using `error_message` would make the message a bit nicer

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 5]
```diff
@@ -0,0 +1,541 @@
...
+                )
+
+        self._gen_val.verify_4xx_response(invalid_token_q)
+
+        if self._test_missing_scope:
+            no_scope_q = self._gen_val.query_missing_scope(**query_kwargs)
+            with self._scenario.check(
+                "Search operational intent references with missing scope", self._pid
+            ) as check:
+                if no_scope_q.status_code != 401:
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:456]: 403?

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+                )
+
+    def _verify_oir_mutation(self):
+        op = OPERATIONS[OperationID.UpdateOperationalIntentReference]
+        new_params = PutOperationalIntentReferenceParameters(**self._oir_params)
+        updated_volume = new_params.extents[0]
+        new_end = updated_volume.time_end.value.datetime - timedelta(seconds=10)
+        updated_volume.time_end = Time(value=StringBasedDateTime(new_end))
+        new_params.extents = [updated_volume]
+        # TODO This feels weird, but is required by the current version of the DSS -> confirm this is as expected
```
[the remark this answers is not in the document]
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py]: Outdated

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md [hunk 7]
```diff
@@ -173,6 +173,138 @@ it is in violation of **[astm.f3548.v21.DSS0210,A2-7-2,7](../../../../../require
 If the DSS does not allow searching for subscriptions when valid credentials are presented,
 it is in violation of **[astm.f3548.v21.DSS0005,5](../../../../../requirements/astm/f3548/v21.md)**.
 
+### Operational intents endpoints authentication test step
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:176]: Makes sense: though, may I suggest that we keep that for a follow-up PR that would address this for both subscriptions and OIRs? 

Currently the subscriptions use the same approach, and it probably makes sense to fix both of them at once.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md (hunk 3 above)
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:178]: I'm not sure: although the check has the same name, they don't point to the same requirement.

One is for `DSS0005,5` (about the subscription API) and the other is for `DSS0005,1` (Op intent refs)

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py [hunk 8]
```diff
@@ -37,7 +41,7 @@ class AuthenticationValidation(TestScenario):
     """
 
     SUB_TYPE = register_resource_type(
-        380, "Subscription, Operational Entity Id, Constraint"
+        382, "Subscription, Operational Entity Id, Constraint"
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py]: Nope, that's a fluke

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py (hunk 1 above)
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py:201]: No, I'd rather add this to my cleanup backlog

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md (hunk 3 above)
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.md:178]: Indeed makes sense.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py (hunk 1 above)
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/authentication_validation.py:201]: Actually: as discussed, we'll rebase on top of #535 to access the validation fragments

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+        self, check: PendingCheck, creation_q: fetch.Query
+    ):
+        try:
+            oir, sanity_check = self._dss.get_op_intent_reference(self._test_id)
+            self._scenario.record_query(sanity_check)
+            if (
+                abs(
+                    oir.time_end.value.datetime
+                    - self._current_oir.time_end.value.datetime
+                ).total_seconds()
```
[the remark this answers is not in the document]
Comment (Shastick) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:527]: `abs()` on a `timedelta`, to my knowledge, produces the expected result: it has been used in some other scenarios such as `monitoring/uss_qualifier/scenarios/astm/utm/dss/subscription_simple.py`

I quickly checked the [python doc](https://docs.python.org/3/library/datetime.html) and that seems correct

**Review by Shastick — COMMENTED** [review]
First batch of updates.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+
+        self._gen_val.verify_4xx_response(invalid_token_q)
+
+        # Valid credentials but missing scope:
+        if self._test_missing_scope:
+            no_scope_q = self._gen_val.query_missing_scope(**query_kwargs)
+            with self._scenario.check(
+                "Create operational intent reference with missing scope", self._pid
+            ) as check:
+                if no_scope_q.status_code != 401:
```
[the remark this answers is not in the document]
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:137]: Ah OK - actually IMO the DSS implementation is wrong then, see the OpenAPI definition:

```yaml
        '403':
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ErrorResponse'
          description: >-
            The access token was decoded successfully but did not include a
            scope appropriate to this endpoint.
```

But then that's not within the scope of this PR. We should probably open an issue on the DSS repo though.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 11]
```diff
@@ -0,0 +1,559 @@
...
+        self._gen_val.verify_4xx_response(invalid_token_q)
+
+        if self._test_missing_scope:
+            no_scope_q = self._gen_val.query_missing_scope(**query_kwargs)
+            with self._scenario.check(
+                "Mutate operational intent reference with missing scope", self._pid
+            ) as check:
+                if no_scope_q.status_code != 401:
+                    check.record_failed(
+                        summary=f"Expected 403, got {no_scope_q.status_code}",
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:315]: Comment mismatch the actual error.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py [hunk 12]
```diff
@@ -0,0 +1,559 @@
...
+        self._gen_val.verify_4xx_response(invalid_token_q)
+
+        if self._test_missing_scope:
+            no_scope_q = self._gen_val.query_missing_scope(**query_kwargs)
+            with self._scenario.check(
+                "Delete operational intent reference with missing scope", self._pid
+            ) as check:
+                if no_scope_q.status_code != 401:
+                    check.record_failed(
+                        summary=f"Expected 403, got {no_scope_q.status_code}",
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/authentication/oir_api_validator.py:403]: Comment mismatch code above.

File: monitoring/uss_qualifier/scenarios/astm/utm/dss/fragments/oir/crud/create_format.md [hunk 13]
```diff
@@ -0,0 +1,9 @@
+# Update operational intent reference response format test step fragment
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/utm/dss/fragments/oir/crud/create_format.md:1]:
```suggestion
# Create operational intent reference response format test step fragment
```

**Review by mickmis — APPROVED** [review]
LGTM - please fix the remaining 403 though before merging

**Pull request merged**